error.go

```diff
@@ -292,6 +309,19 @@ func wrapIfUncoded(err error) error {
 	return NewError(CodeUnknown, maybeCodedErr)
 }
 
+// wrapIfEOF is used for abnormal EOF conditions, like EOF before response
+// headers were received. When this happens, we want errors.Is(err, io.EOF) to
+// return false (since we have *many* tests for io.EOF, which otherwise
+// indicates a normal end of stream). Instead, the error will return true with
+// errors.Is(err, io.ErrUnexpectedEOF), to make it more clear that the EOF is
+// an abnormal end of the operation.
+func wrapIfEOF(err error) error {
+	if errors.Is(err, io.EOF) {
+		return &unexpectedEOFError{err}
```

Comment:
No, that doesn't work at all. In fact, that's what the code _already does_. If you leave that wrapping and remove this one, the tests will fail with a mix of "unimplemented" and "unknown" error codes, instead of "unavailable".

I wrote long-ish comments trying to explain why this is needed. Were they unclear?

Reply:
Well, that's basically what the other PR did -- it replaced the error instead of wrapping it.

I also now realize that this PR that "wraps" isn't _really_ wrapping it, at least not properly since it doesn't provide an `Unwrap() error` method. And adding such a method causes the tests to fail... So I'll just stick with what was in #776.

Reply:
Sounds good, but for consistency with the other error handlers the `wrap` was meant purely to wrap it in a connect error code. All the other `wrap...` functions call `NewError` so this should wrap in one too. It avoids other wrappers looking at it as each one checks if its already wrapped before processing. It has nothing to do with the `Unwrap() error` interface (I mean `NewError` does implement it but thats not the point). My comments were only to enforce the consistency of the current error handling.